src/procevent.c

```diff
@@ -1207,6 +1221,10 @@ static int procevent_init(void) /* {{{ */
   ring.tail = 0;
   ring.maxLen = buffer_length;
   ring.buffer = (cdtime_t **)calloc(buffer_length, sizeof(cdtime_t *));
+  if (ring.buffer == NULL) {
+    ERROR("procevent plugin: Allocating buffer failed.");
+    return ENOMEM;
+  }
```

Comment:
I think it's bad style to call `exit()` from outside the `main` function. I'd rather remove `smalloc()` and `scalloc()` and add error handling to the few places they're used in.

Reply:
> FYI, the main reason for adding this is that I ran into the problem that `buffer_length` was zero. This check would have caught that situation.

I think scalloc() would have caught and logged that too (before exiting)?

Reply:
I think that we should strive to be resilient even under memory pressure. I realize that to get there we would actually have to test collectd's behavior, e.g. by implementing fuzz testing and injecting (random) alloc failures. We don't have that now, but if the code is at least mostly resilient we have a chance of getting there eventually.

Reply:
Generating the errors is the trivial part. Actually verifying that something meaningful is done as result is the hard part.

Because any allocations can fail, it's not enough to test just each allocation failure in isolation, but _also_ their combinations.  If it's done properly, one can easily end up with many times larger amount of code for validating the error handling, than code for the plugin itself.  Which also needs to be reviewed and maintained.  And all that for something that is not even supposed to happen! [1]

[1] If program runs out of memory, most likely system / container / process limits are improperly configured, or program leaks.  In which case there's not much point in trying to limp further, as things will only get worse.   If config is wrong, it should be fixed.  If program leaks, restarting is the only thing helping, before leak is fixed.

Only case where limping along slightly longer may make some sense, is some other program temporarily taking just enough memory that the system is on edge of not working, but can still somehow function (some allocs fail, but **not** all), and if restart would also require temporarily more memory, or be too slow in such situation.   I think on systems like Linux with overcommit enabled, that' too much of a corner-case though.

=> I would rather fail fast, and spend testing on making sure that restarts work reliably.